src/mlpack/core/dists/discrete_distribution.cpp

```diff
@@ -20,22 +20,29 @@ using namespace mlpack::distribution;
  */
 arma::vec DiscreteDistribution::Random() const
 {
-  // Generate a random number.
-  double randObs = math::Random();
-  arma::vec result(1);
+  size_t dimension = probabilities.size();
+  arma::vec result(dimension);
 
-  double sumProb = 0;
-  for (size_t obs = 0; obs < probabilities.n_elem; obs++)
+  for (size_t d = 0; d < dimension; d++)
   {
-    if ((sumProb += probabilities[obs]) >= randObs)
+    // Generate a random number.
+    double randObs = math::Random();
+
+    double sumProb = 0;
+
+    for (size_t obs = 0; obs < probabilities[d].n_elem; obs++)
     {
-      result[0] = obs;
-      return result;
+      if ((sumProb += probabilities[d][obs]) >= randObs)
+      {
+        result[d] = obs;
+        break;
+      }
     }
-  }
 
-  // This shouldn't happen.
-  result[0] = probabilities.n_elem - 1;
+    if (sumProb >= randObs != true)
```

Comment:
Why the `!= true`?  That seems strange.  This should be just `if (sumProb >= randObs)`, I think.